sdk/data/azcosmos/cosmos_location_cache.go

```diff
@@ -0,0 +1,303 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azcosmos
+
+import (
+	"net/url"
+	"sync"
+	"time"
+)
+
+const defaultExpirationTime time.Duration = time.Minute * 5
+
+const (
+	none opType = iota
+	read
+	write
+	all
+)
+
+type opType int
+
+type locationUnavailabilityInfo struct {
+	lastCheckTime  time.Time
+	unavailableOps opType
+}
+
+type dbAcctLocationsInfo struct {
+	prefLocations              []string
+	availWriteLocations        []string
+	availReadLocations         []string
+	availWriteEndptsByLocation map[string]url.URL
+	availReadEndptsByLocation  map[string]url.URL
+	writeEndpts                []url.URL
+	readEndpts                 []url.URL
+}
+
+// AcctRegion represents a Azure Cosmos DB Database Account in a specific region.
+type AcctRegion struct {
+	// The name of the database account location in the Azure Cosmos DB service.
+	name string
+	// The URL of the database account location in the Azure Cosmos DB service.
+	endpoint string
+}
+
+// AcctProperties represents a container for Azure Cosmos DB databases.
+type AcctProperties struct {
+	// The list of readable regions of this database account from the Azure Cosmos DB service.
+	readRegions []AcctRegion
+	// The list of writable regions of this database account from the Azure Cosmos DB service.
+	writeRegions []AcctRegion
+	// Indicates whether multiple write locations are enabled for this database account.
+	enableMultipleWriteLocations bool
+}
+
+type locationCache struct {
+	locationInfo              dbAcctLocationsInfo
+	defaultEndpt              url.URL
+	enableEndptDiscovery      bool
+	useMultipleWriteLocations bool
+	rwMutex                           sync.RWMutex
+	locationUnavailabilityInfoMap     map[url.URL]locationUnavailabilityInfo
+	mapMutex                          sync.RWMutex
+	lastUpdateTime                    time.Time
+	enableMultipleWriteLocations      bool
+	unavailableLocationExpirationTime time.Duration
+}
+
+func newLocationCache(prefLocations []string, defaultEndpt url.URL) *locationCache {
+	return &locationCache{
+		defaultEndpt:                      defaultEndpt,
+		locationInfo:                      *newDbAcctLocationsInfo(prefLocations, defaultEndpt),
+		locationUnavailabilityInfoMap:     make(map[url.URL]locationUnavailabilityInfo),
+		unavailableLocationExpirationTime: defaultExpirationTime,
+	}
+}
+
+func (lc *locationCache) update(writeLocations []AcctRegion, readLocations []AcctRegion, prefList []string, enableMultipleWriteLocations bool) error {
+	lc.rwMutex.RLock()
+	nextLoc := copyDbAcctLocationsInfo(lc.locationInfo)
+	if prefList != nil {
```

Comment:
`nil` comparison will work if the parameter is a pointer, like `*[]string`

Reply:
As far as I understand, the zero/default value for slice is nil, so it works fine as is in that regard. Passing parameters as pointers are typically only needed if they are going to be modified, correct? The only benefit I can think of is that, for large slices, it may be more efficient to pass the pointer vs. pass-by-value (i.e. copy the slice).

Reply:
If nil works (and we can verify with tests) sure. I was under the impression (wrongly) that `nil` only worked with pointers and optional parameters should use pointers. The pattern used for Go SDKs' public API is for all optional parameters to be pointers.

Reply:
According to https://azure.github.io/azure-sdk/golang_introduction.html#optional-parameters, it seems that is best to create a struct type representing a set of optional parameters to provide. I believe this is only for the public-facing API, so in this case, I think we can leave as is. The tests already cover the functionality, including passing nil directly.